cpp/comms/src/cuML_std_comms_impl.cpp

```diff
@@ -0,0 +1,417 @@
+    case MLCommon::cumlCommunicator::FLOAT:
+      return sizeof(float);
+    case MLCommon::cumlCommunicator::DOUBLE:
+      return sizeof(double);
+  }
+}
+
+ncclDataType_t getNCCLDatatype(
+  const cumlStdCommunicator_impl::datatype_t datatype) {
+  switch (datatype) {
+    case MLCommon::cumlCommunicator::CHAR:
+      return ncclChar;
+    case MLCommon::cumlCommunicator::UINT8:
+      return ncclUint8;
+    case MLCommon::cumlCommunicator::INT:
+      return ncclInt;
+    case MLCommon::cumlCommunicator::UINT:
+      return ncclUint32;
+    case MLCommon::cumlCommunicator::INT64:
+      return ncclInt64;
+    case MLCommon::cumlCommunicator::UINT64:
+      return ncclUint64;
+    case MLCommon::cumlCommunicator::FLOAT:
+      return ncclFloat;
+    case MLCommon::cumlCommunicator::DOUBLE:
+      return ncclDouble;
+  }
+}
+
+ncclRedOp_t getNCCLOp(const cumlStdCommunicator_impl::op_t op) {
+  switch (op) {
+    case MLCommon::cumlCommunicator::SUM:
+      return ncclSum;
+    case MLCommon::cumlCommunicator::PROD:
+      return ncclProd;
+    case MLCommon::cumlCommunicator::MIN:
+      return ncclMin;
+    case MLCommon::cumlCommunicator::MAX:
+      return ncclMax;
+  }
+}
+}  // namespace
+
+
+bool ucx_enabled(){
+  return UCX_ENABLED;
+}
+
+/**
+ * @brief Underlying comms, like NCCL and UCX, should be initialized and ready for use,
+ * and maintained, outside of the cuML Comms lifecycle. This allows us to decouple the
+ * ownership of the actual comms from cuml so that they can also be used outside of cuml.
+ *
+ * For instance, nccl-py can be used to bootstrap a ncclComm_t before it is
+ * used to construct a cuml comms instance. UCX endpoints can be bootstrapped
+ * in Python using ucx-py, before being used to construct a cuML comms instance.
+ */
+#ifdef WITH_UCX
+void inject_comms(cumlHandle &handle, ncclComm_t comm, ucp_worker_h ucp_worker,
+                  std::shared_ptr<ucp_ep_h *> eps, int size, int rank) {
+  auto communicator = std::make_shared<MLCommon::cumlCommunicator>(
+    std::unique_ptr<MLCommon::cumlCommunicator_iface>(
+      new cumlStdCommunicator_impl(comm, ucp_worker, eps, size, rank)));
+  handle.getImpl().setCommunicator(communicator);
+}
+#endif
+
+void inject_comms(cumlHandle &handle, ncclComm_t comm, int size, int rank) {
+  auto communicator = std::make_shared<MLCommon::cumlCommunicator>(
+    std::unique_ptr<MLCommon::cumlCommunicator_iface>(
+      new cumlStdCommunicator_impl(comm, size, rank)));
+  handle.getImpl().setCommunicator(communicator);
+}
+
+void inject_comms_py_coll(cumlHandle *handle, ncclComm_t comm, int size,
+                          int rank) {
+  inject_comms(*handle, comm, size, rank);
+}
+
+void inject_comms_py(ML::cumlHandle *handle, ncclComm_t comm, void *ucp_worker,
+                     void *eps, int size, int rank) {
+
+
+#ifdef WITH_UCX
+  ucp_worker_print_info((ucp_worker_h)ucp_worker, stdout);
+
+  // Make shared_ptr for endpoints
+  std::shared_ptr<ucp_ep_h *> eps_sp =
```

Comment:
To my understanding the caller of `inject_comms_py` owns `eps` and `ucp_worker` and is therefore responsible for releasing the backing store of `eps` and `ucp_worker` as well as the resources the entries of `eps` and `ucp_worker`  point to. Therefore I am not sure why the introduction of this shared_ptr structure is needed. The interface of `inject_comms` could be changed to accept a raw pointer and `inject_comms_py` simply does the necessary reinterpret casts.

Reply:
We discussed this offline and, given my limited knowledge of Cython, I think I might have a better solution to try.